### PR TITLE
F5 LTM

### PR DESCRIPTION
**Pull request by adaniels21487**
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [X] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [X] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

This module performs discovery and polling of LTM objects on F5 BigIP devices.

It contains the following features:
- Discovery of Virtual servers, Pools and Pool Members.
- Collection and Graphing of the following metrics across Virtual Servers and Pool Members:
  - Bytes In/Out, Packets In/Out, Connections
- Pagination and Searching of tables via bootgrid.
- Alerting on the following conditions:
  - Virtual Server Down - Critical (2)
  - Pool does not have the minimum required members - Critical (2)
  - Pool Member down - warning (1)

### CODE REVIEW

**Comment by laf** [comment]
I have some suggestions to make here - nothing major, but give me a couple of days maybe.

File: LibreNMS/Component.php [hunk 1]
```diff
@@ -41,6 +41,29 @@ class Component
...
+        if (is_null($device_id)) {
+            // SELECT type, count(*) as count FROM component GROUP BY type
+            $SQL = "SELECT `type` as `name`, count(*) as count FROM `component` GROUP BY `type`";
+            $rows = dbFetchRows($SQL, array());
+        } else {
+            $SQL = "SELECT `type` as `name`, count(*) as count FROM `component` WHERE `device_id` = ? GROUP BY `type`";
+            $rows = dbFetchRows($SQL, array($device_id));
+        }
+
+        if (!isset($rows)) {
```
Comment (laf) [inline LibreNMS/Component.php]: Why not just `if (isset($rows)) {` do you 'We found some, lets re-process to make more accessible' bit then outside of the `if` block just `return false;` to cover the if not matching.

Comment (laf) [inline LibreNMS/Component.php]:
```php
if (isset($rows)) {
    $result = array();
    foreach ($rows as $value) {
        $result[$value['name']] = $value['count'];
    }
    return $result;
}
return false;
```

File: html/pages/device.inc.php [hunk 2]
```diff
@@ -170,6 +173,18 @@
             }
         }
 
+        if ($device['os'] == 'f5') {
```
Comment (laf) [inline html/pages/device.inc.php]: Don't need this check really, only f5 devices would have the data so the isset check below should be adequate no?

File: html/pages/device/loadbalancer.inc.php [hunk 3]
```diff
@@ -6,12 +6,11 @@
...
-// Cisco ACE
-$type_text['loadbalancer_rservers'] = 'Rservers';
-$type_text['loadbalancer_vservers'] = 'Serverfarms';
-
-// Citrix Netscaler
-$type_text['netscaler_vsvr'] = 'VServers';
+$type_text['loadbalancer_rservers'] = 'Rservers';       // Cisco ACE
+$type_text['loadbalancer_vservers'] = 'Serverfarms';    // Cisco ACE
+$type_text['netscaler_vsvr'] = 'VServers';              // Citrix Netscaler
+$type_text['ltm-vs'] = 'LTM Virtual Servers';           // F5 BigIP
```
Comment (laf) [inline html/pages/device/loadbalancer.inc.php]: can we change these to be ltm_vs and ltm_pool to be consistent with the above.

File: includes/defaults.inc.php [hunk 4]
```diff
@@ -705,6 +705,7 @@
 $config['poller_modules']['cisco-otv']                   = 1;
 $config['poller_modules']['ntp']                         = 1;
 $config['poller_modules']['services']                    = 1;
+$config['poller_modules']['loadbalancers']               = 1;
```
Comment (laf) [inline includes/defaults.inc.php]: We now have the ability to enable / disable these on a per os basis so set them to 0 here and enable them in the os. See `$os = 'linux';` in definitions.inc.php to see.

File: includes/defaults.inc.php [hunk 5]
```diff
@@ -742,6 +743,7 @@
 $config['discovery_modules']['stp']            = 1;
 $config['discovery_modules']['cisco-otv']      = 1;
 $config['discovery_modules']['ntp']            = 1;
+$config['discovery_modules']['loadbalancers']  = 1;
```
Comment (laf) [inline includes/defaults.inc.php]: We now have the ability to enable / disable these on a per os basis so set them to 0 here and enable them in the os. See `$os = 'linux';` in definitions.inc.php to see.

File: includes/discovery/loadbalancers/f5-ltm.inc.php [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                // All is good.
+                $result['status'] = 0;
+                $result['error'] = '';
+            }
+        }
+
+        // Do we have any results
+        if (count($result) > 0) {
+            // Let's log some debugging
+            d_echo("\n\n".$result['type'].": ".$result['label']."\n");
```
Comment (laf) [inline includes/discovery/loadbalancers/f5-ltm.inc.php]: All of these d_echo()'s (in all the code below as well) is a bit much, can you condense it down or remove unnecessary debug info

File: includes/polling/loadbalancers/f5-ltm.inc.php [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            $fields = array(
+                'pktsin' => $ltmVirtualServStatEntry['1.3.6.1.4.1.3375.2.2.10.2.3.1.6.'.$UID],
+                'pktsout' => $ltmVirtualServStatEntry['1.3.6.1.4.1.3375.2.2.10.2.3.1.8.'.$UID],
+                'bytesin' => $ltmVirtualServStatEntry['1.3.6.1.4.1.3375.2.2.10.2.3.1.7.'.$UID],
+                'bytesout' => $ltmVirtualServStatEntry['1.3.6.1.4.1.3375.2.2.10.2.3.1.9.'.$UID],
+                'totconns' => $ltmVirtualServStatEntry['1.3.6.1.4.1.3375.2.2.10.2.3.1.11.'.$UID],
+            );
+
+            // Let's print some debugging info.
+            d_echo("\n\nComponent: ".$key."\n");
```
Comment (laf) [inline includes/polling/loadbalancers/f5-ltm.inc.php]: Same here for d_echo use.

File: sql-schema/152.sql [hunk 8]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `component_prefs` CHANGE `value` `value` VARCHAR(4000) NOT NULL COMMENT 'Value for the Component';
```
Comment (laf) [inline sql-schema/152.sql]: Does it need to be 4000? Is a text field not better?

File: LibreNMS/Component.php (hunk 1 above)
Comment (laf) [inline LibreNMS/Component.php]: Think you need to push the commit :)

File: html/pages/device.inc.php (hunk 2 above)
Comment (adaniels21487) [inline html/pages/device.inc.php]: Done, not yet pushed.

File: html/pages/device/loadbalancer.inc.php (hunk 3 above)
Comment (adaniels21487) [inline html/pages/device/loadbalancer.inc.php]: Done, not yet pushed.

File: includes/defaults.inc.php (hunk 5 above)
Comment (adaniels21487) [inline includes/defaults.inc.php]: Done, not yet pushed.

File: includes/defaults.inc.php (hunk 4 above)
Comment (adaniels21487) [inline includes/defaults.inc.php]: Done, not yet pushed.

File: sql-schema/152.sql (hunk 8 above)
Comment (adaniels21487) [inline sql-schema/152.sql]: Done, not yet pushed.

File: includes/polling/loadbalancers/f5-ltm.inc.php (hunk 7 above)
Comment (adaniels21487) [inline includes/polling/loadbalancers/f5-ltm.inc.php]: I have removed all the counter data as the RRD debugs already show these, Not yet pushed.

File: includes/discovery/loadbalancers/f5-ltm.inc.php (hunk 6 above)
Comment (adaniels21487) [inline includes/discovery/loadbalancers/f5-ltm.inc.php]: Trimmed. not yet pushed.

**Comment by laf** [comment]
Thanks as always @adaniels21487 for contributing 👍

**Comment by laf** [comment]
@adaniels21487 sorry we've had to revert this. `$ltmVirtualServStatEntry = snmpwalk_array_num($device, '.1.3.6.1.4.1.3375.2.2.10.2.3.1', 0);` causes a huge snmpwalk to take place and for one user this caused an increase in poller times over 5 minutes.

I've got debug output if you need to see it but from what I can see this seems to walk an entire OID tree of over 50k lines back before the process was killed manually.

**Comment by awlx** [comment]
To add some facts about the system which is polled. It has like 2000 VIPs configured and around the same number of Pools. 

But I have seen bigger deployments in Shared Hosting envs.

**Comment by adaniels21487** [comment]
Thanks for the info @awlx
It sounds like we have two options.
1. Leave the PR as is and disable the discovery module on devices with too many VS's / pools.
2. Change the PR so modules default off and users can manually enable them if desired.

I would have thought having this many VS's / pools would be the exception and not the rule, leaning toward option 1, but I'm easy either way.

Thanks Aaron

**Comment by laf** [comment]
@adaniels21487 it's not discovery but polling that is causing the issue, take a look at my response above. That walk is grabbing lots of data which seems unnecessary. I'm sure it can be fixed easily but I just don't have the relevant kit or walk to be able to simulate and help.

**Comment by adaniels21487** [comment]
No worries. I am away for Christmas, will take a look when I get back.

**Comment by adaniels21487** [comment]
My wife thinks I'm packing :)

@awlx I have made some changes and re-submitted this in #5205 
Could you please test and advise if this resolves the polling issue?

Thanks,
Aaron

**Comment by awlx** [comment]
Testing the new code. So far everything looks good just a bit increased Poller Times. But this was expected. 

http://i.imgur.com/unC01fs.png

**Comment by awlx** [comment]
I just saw that the "Loadbalancer" tap does not show up on the two LBs which have many VIPs. But on the test-lb it shows up.

**Comment by awlx** [comment]
Ok, I found out that the discovery module was failing because of the fanspeed issue which is already fixed in master. 

Now after I fixed that discovery hangs when discovering the new OIDs.

**Comment by awlx** [comment]
When discovery.php stops working the only thing I see is that it runs at 100% on one core. And strace gives the following output:

> strace -f -p 58229
> Process 58229 attached
> 
> brk(0x6344000)                          = 0x6344000
> brk(0x6384000)                          = 0x6384000
> brk(0x63c4000)                          = 0x63c4000
> brk(0x6404000)                          = 0x6404000
> brk(0x6444000)                          = 0x6444000
> brk(0x6484000)                          = 0x6484000
> brk(0x64c4000)                          = 0x64c4000
> brk(0x6504000)                          = 0x6504000
> brk(0x6544000)                          = 0x6544000
> brk(0x6584000)                          = 0x6584000
> brk(0x65c4000)                          = 0x65c4000
> brk(0x6604000)                          = 0x6604000
> brk(0x6644000)                          = 0x6644000
> brk(0x6684000)                          = 0x6684000
> brk(0x66c4000)                          = 0x66c4000
>

**Comment by awlx** [comment]
And now after the running discovery.php poller.php dies. WIth zend_mm_heap corrupted

**Comment by laf** [comment]
I think we'll hold off with merging the f5 code for now.

**Comment by adaniels21487** [comment]
Hi @awlx 
The discovery module was not changed between the previous PR and this one. Did this issue occur last time?

Can you please send a pastebin link to the discovery debug output so we can see where it dies.

**Comment by awlx** [comment]
Discovery runs through it just takes 20mins for a run. After discovery went through each time poller tries to run the poller dies with zend_mm_heap corrupted. I think it's just way too much data.

**Comment by adaniels21487** [comment]
Can you please send me a pastebin link for the poller debug.

**Comment by awlx** [comment]
There is no output just "zend_mm_heap corrupted" when I start it.

**Comment by adaniels21487** [comment]
There should be output before the F5 module is loaded, lots if it is a debug, if you dont see this then there are other issues at play.

**Comment by awlx** [comment]
I raised the memory for PHP to 12GB now.

This is what happens now:
>> Runtime for poller module 'services': 0.0000 seconds
#### Unload poller module services ####

RRD[update /opt/librenms/rrd/lb01.company/poller-perf-services.rrd N:4.5061111450195E-5]
Modules status: Global- OS+ Device  
#### Load poller module loadbalancers ####
SQL[SELECT `C`.`id`,`C`.`device_id`,`C`.`type`,`C`.`label`,`C`.`status`,`C`.`disabled`,`C`.`ignore`,`C`.`error`,`CP`.`attribute`,`CP`.`value` FROM `component` as `C` LEFT JOIN `component_prefs` as `CP` on `C`.`id`=`CP`.`component` WHERE  ( `disabled` = '0' AND `ignore` = '0' AND `device_id` = '6' )] 
Segmentation fault (core dumped)

**Comment by adaniels21487** [comment]
Can you run that uery directly in mysql:

```
SELECT C.id,C.device_id,C.type,C.label,C.status,C.disabled,C.ignore,C.error,CP.attribute,CP.value FROM component as C LEFT JOIN component_prefs as CP on C.id=CP.component WHERE ( disabled = '0' AND ignore = '0' AND device_id = '6' )
```

**Comment by awlx** [comment]
So it seems that every 5th try runs here is the output of poller:
https://cloud.awlnx.space/owncloud/index.php/s/6RQUmUTnhkdgJpl

**Comment by adaniels21487** [comment]
There seems to be an awful lot of DB updates.
These look like issues in the component class, I will investigate.

Any chance you can remove all F5 components and re-discover?

```
DELETE from component where type like 'f5-%';
```

**Comment by awlx** [comment]
I did as you said. Runtime was now much faster:
1 devices polled in 14.87 secs

**Comment by awlx** [comment]
I did as you said. Rediscovering now. This will take a while.

**Comment by awlx** [comment]
Discovery went through:
2017-01-03 17:30:31 - 1 devices discovered in 986.2 secs

Poller:
Adds the components again.
 1 devices polled in 73.51 secs

2nd run:
Poller updates all the components again.
1 devices polled in 35.20 secs

3rd run:
Poller updates less components
1 devices polled in 18.99 secs

**Comment by awlx** [comment]
I let it run the whole night and it looks good for now. Polling times for the loadbalancers are between 30 - 45s.

**Comment by laf** [comment]
@adaniels21487 Do we need to consider doing something to clear up the DB for people in case others have this issue?

**Comment by awlx** [comment]
After some time passed, it happens again:
SQL[SELECT `C`.`id`,`C`.`device_id`,`C`.`type`,`C`.`label`,`C`.`status`,`C`.`disabled`,`C`.`ignore`,`C`.`error`,`CP`.`attribute`,`CP`.`value` FROM `component` as `C` LEFT JOIN `component_prefs` as `CP` on `C`.`id`=`CP`.`component` WHERE  ( `device_id` = '6' )] 
zend_mm_heap corrupted

**Comment by adaniels21487** [comment]
So this query is run at the end of the cycle, it is run to compare the pre-poll and post-poll status , to see if it has changed.

Can you please let me know how many items are being pulled:

```
SELECT `type` as `name`, count(*) as count FROM `component` WHERE `device_id` = 6 GROUP BY `type`;
```

I am guessing this error is something todo with the MySQL PDO module. A big of googling found this:
https://bugs.php.net/bug.php?id=53322
Is it possible to upgrade/downgrade your PHP and see if this issue persists?

**Comment by laf** [comment]
Can we move this conversation to #5205 please as it's more relevant there now.

**Pull request merged**